soc/arm/renesas_smartbond/da1469x/soc.c

```diff
@@ -42,6 +53,50 @@ void sys_arch_reboot(int type)
 	NVIC_SystemReset();
 }
 
+void
+z_smartbond_vdd_level_set(uint8_t vdd_level)
+{
+	POWER_CTRL_REG_SET(VDD_LEVEL, vdd_level);
+}
+
+static void z_smartbond_configure_power_rails(void)
+{
+	POWER_CTRL_REG_SET(LDO_3V0_MODE, 3);
+	POWER_CTRL_REG_SET(LDO_3V0_REF, 1);
+	POWER_CTRL_REG_SET(V30_LEVEL, V30_LEVEL_3V00);
+	POWER_CTRL_REG_SET(LDO_3V0_RET_ENABLE_ACTIVE, 0);
+	POWER_CTRL_REG_SET(LDO_3V0_RET_ENABLE_SLEEP, 1);
+	POWER_CTRL_REG_SET(CLAMP_3V0_VBAT_ENABLE, 0);
+
+	/* Enable V18, use LDO_IO in active mode, disable in sleep */
+	POWER_CTRL_REG_SET(V18_LEVEL, V18_LEVEL_1V8);
+	POWER_CTRL_REG_SET(LDO_1V8_RET_ENABLE_ACTIVE, 0);
+	POWER_CTRL_REG_SET(LDO_1V8_RET_ENABLE_SLEEP, 0);
+	POWER_CTRL_REG_SET(LDO_1V8_ENABLE, 1);
+
+	/* Enable V18P, use LDO_IO2 in active mode and LDO_IO_RET2 in sleep */
+	POWER_CTRL_REG_SET(LDO_1V8P_RET_ENABLE_ACTIVE, 0);
+	POWER_CTRL_REG_SET(LDO_1V8P_RET_ENABLE_SLEEP, 1);
+	POWER_CTRL_REG_SET(LDO_1V8P_ENABLE, 1);
+
+	/* Configure VDD to 1.2V if PLL is enabled, 0.9V otherwise. */
+	if (DT_NODE_HAS_STATUS(DT_NODELABEL(pll), okay)) {
```

Comment:
The VDD level configuration should be removed here. This is already handled by the clock control driver. Another reason is that changing the VDD level should always be accompanied with updating the flash controller's pipe clock delay. Otherwise the controller will become unstable and this should end up raising ARM exceptions. The issue can be reproduced when disabling PLL (by default should be enabled in da1469x_dk_pro overlay).

Reply:
@gmarull Added regulator driver that is now used
@ioannis-karachalios soc.c now does not set VDD voltage in init. It is done in regulator devices. Only when clock control requests PLL voltage is increased.

@ioannis-karachalios This code can be tested with regulator shell (preferable with #62190 applied). I noticed that when DCDC is selected for V18 rail and this rail is configured to 1.2V instead of default 1.8V. Rail is no powered during active mode and only is powered when device goes to sleep when LDO_RET is used instead. When rail is set to 1.8V LDO and DCDC work correct in active mode.

Reply:
@kasjer, as for setting VDD voltage, I knew that regulator devices would be introduced and so the existing implementation would no longer be valid. I just mentioned it, so you are aware of VDD update dependencies.  Next week, I will test the code with regulator shell and let you know if what you observe for V18 rail is expected or not.  Thanks for the update.

Reply:
@kasjer as for V18 rail not being able to generate 1.2V, is `DCDC_V18_REG[DCDC_V18_ENABLE_LV]` bit field set? Otherwise, low voltages cannot be generated (default value should be zero) which should also be reflected in the default device tree property values:

```
 v18: V18 {
     ...
    regulator-initial-mode =
	<DA1469X_MODE_DCDC_HIGH_BATTERY_VOLTAGE_ON>;
 };
```

Reply:
@ioannis-karachalios maybe I had something set incorrect. With current code I don't see this anomaly any more. V18 when set to 1.2V works OK with DCDC and LDO in active and sleep mode.

Reply:
Thanks for the update. I just confirmed internally that `DCDC_V18_ENABLE_LV` bit field is not required for V18 rail to generate 1.2V